### PR TITLE
Support for qnx

### PR DESCRIPTION
**Pull request by hyarasi13**
Opencv 4.7.0 support for QNX sdp 7.1(QNXOSG-245)

This PR introduces support for OpenCV 4.7.0 on QNX SDP 7.1.
Development was based on our qnx-ports infrastructure, with references from both internal GitLab repositories and the upstream OpenCV baseline.

Passed tests: 36,625
Failed tests: 98

Test instructions are available in build-files/ports/opencv under qnx-ports.

Tests failed are listed below:
[ FAILED ] 6 tests, listed below:
[ FAILED ] Core_InputOutput.filestorage_base64_basic_read_XML
[ FAILED ] Core_InputOutput.filestorage_base64_basic_read_YAML
[ FAILED ] Core_InputOutput.filestorage_base64_basic_read_JSON
[ FAILED ] Core_InputOutput.filestorage_base64_basic_rw_XML
[ FAILED ] Core_InputOutput.filestorage_base64_basic_rw_YAML
[ FAILED ] Core_InputOutput.filestorage_base64_basic_rw_JSON

[ FAILED ] 85 tests, listed below:
[ FAILED ] Test_Caffe.shared_weights
[ FAILED ] Test_Int8_layers.Convolution2D/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Int8_layers.LeakyReLU/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Int8_layers.ReLU6/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Int8_layers.BatchNorm/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Int8_layers.InnerProduct/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Int8_layers.Eltwise/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Caffe_layers.PReLU/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Caffe_layers.layer_prelu_fc/0, where GetParam() = OCV/CPU
[ FAILED ] Test_Caffe_layers.Convolution2Inputs/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_celu_expanded_OCV_CPU, where GetParam() = (test_celu_expanded, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_elu_OCV_CPU, where GetParam() = (test_elu, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_elu_example_OCV_CPU, where GetParam() = (test_elu_example, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_hardsigmoid_OCV_CPU, where GetParam() = (test_hardsigmoid, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_hardsigmoid_example_OCV_CPU, where GetParam() = (test_hardsigmoid_example, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_hardswish_expanded_OCV_CPU, where GetParam() = (test_hardswish_expanded, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_leakyrelu_OCV_CPU, where GetParam() = (test_leakyrelu, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_leakyrelu_example_OCV_CPU, where GetParam() = (test_leakyrelu_example, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_lrn_OCV_CPU, where GetParam() = (test_lrn, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_selu_OCV_CPU, where GetParam() = (test_selu, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_selu_example_OCV_CPU, where GetParam() = (test_selu_example, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_shrink_hard_OCV_CPU, where GetParam() = (test_shrink_hard, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_shrink_soft_OCV_CPU, where GetParam() = (test_shrink_soft, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_thresholdedrelu_OCV_CPU, where GetParam() = (test_thresholdedrelu, OCV/CPU)
[ FAILED ] Test_ONNX_conformance.Layer_Test/test_thresholdedrelu_example_OCV_CPU, where GetParam() = (test_thresholdedrelu_example, OCV/CPU)
[ FAILED ] Test_ONNX_layers.InstanceNorm/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Clip/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Clip_init/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.MatMul_2d_init/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.MatMul_4d_init/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.MatMul_init_bcast/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Resize/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.DivConst/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Gemm/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Convolution/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_MatMul/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Gemm/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Eltwise/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Eltwise_Scalar/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Eltwise_Broadcast/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_LeakyReLU/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Sigmoid/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_MaxPool/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_AvgPool/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Split/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Pad/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Reshape/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Transpose/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Squeeze/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Unsqueeze/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Resize/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Concat/0, where GetParam() = OCV/CPU
[ FAILED ] Test_ONNX_layers.Quantized_Constant/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.reduce_mean/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.reduce_max/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.eltwise/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_3/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_4/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_7/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_8/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_9/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_12/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm_13/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.batch_norm3D/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.slim_batch_norm/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_max_pool_odd_same/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_0_false/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_1_false/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_2_false/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_3_false/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_1_2_false/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_0_true/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_1_true/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_2_true/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_3_true/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.pooling_reduce_sum_1_2_true/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.deconvolution/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.leaky_relu/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.BiasAdd/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.ExpandDims/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.relu6/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.keras_mobilenet_head/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.resize_bilinear_down/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.clip_by_value/0, where GetParam() = OCV/CPU
[ FAILED ] Test_TensorFlow_layers.tf2_prelu/0, where GetParam() = OCV/CPU

[ FAILED ] 3 tests, listed below:
[ FAILED ] Highgui_GUI.regression
[ FAILED ] Highgui_GUI.trackbar_unsafe
[ FAILED ] Highgui_GUI.trackbar

[ FAILED ] 2 tests, listed below:
[ FAILED ] Imgproc_ColorBayerVNG.regression
[ FAILED ] Imgproc_ColorBayerVNG_Strict.regression

[ FAILED ] 2 tests, listed below:
[ FAILED ] Photo_MergeDebevec.regression
[ FAILED ] Photo_CalibrateDebevec.regression

skipped due to memory crash (SIGBUS error: unaligned memory access)

Caffe:
Test_Caffe_layers.Convolution/0 
Test_Caffe_layers.DeConvolution/0
Test_Caffe_layers.InnerProduct/0 
Test_Caffe_layers.BatchNorm/0
Test_Caffe_layers.Concat/0 
Int8:
Test_int8_layers.Concat/0 
ONNX:
Test_ONNX_layers.MatMul_3d_init/0 
Test_ONNX_layers.MatMul_init_2/0 
Test_ONNX_layers.LSTM_Activations/0 
TensorFlow:
Test_TensorFlow_layers.tf_reshape_nhwc/0

### CODE REVIEW

File: modules/core/include/opencv2/core/utils/filesystem.private.hpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// Some modifications 2025 by BlackBerry Limited.
```
Comment (dchordia) [inline modules/core/include/opencv2/core/utils/filesystem.private.hpp]: Can be removed

File: modules/ts/src/ts.cpp [hunk 2]
```diff
@@ -1033,7 +1035,12 @@ static std::string findData(const std::string& relative_path, bool required, boo
         }
     }
 #ifdef OPENCV_TEST_DATA_INSTALL_PATH
+#ifdef __QNX__
```
Comment (dchordia) [inline modules/ts/src/ts.cpp]: Why add this? Usual `datapath` does seem to install correctly, please check

File: modules/objdetect/test/test_qrcode_encode.cpp [hunk 3]
```diff
@@ -264,7 +264,8 @@ TEST(Objdetect_QRCode_Encode_Decode, regression)
                 int true_capacity = establishCapacity(mode, version, cur_capacity);
 
                 std::string input_info = symbol_set;
-                std::random_shuffle(input_info.begin(),input_info.end());
+                //std::random_shuffle(input_info.begin(),input_info.end());
```
Comment (dchordia) [inline modules/objdetect/test/test_qrcode_encode.cpp]: What problem does using `random_shuffle` cause?

**Comment by dchordia** [comment]
Also, do add package install options, similar to `qnx-4.9.0` has

File: modules/objdetect/test/test_qrcode_encode.cpp (hunk 3 above)
Comment (hyarasi13) [reply modules/objdetect/test/test_qrcode_encode.cpp]: std::random_shuffle was deprecated in C++14 and completely removed in C++17 which is causing error as 'random_shuffle' is not a member of 'std'
